File: src/main/java/bad/robot/radiate/teamcity/BuildTypes.java
package bad.robot.radiate.teamcity;

import com.google.gson.annotations.SerializedName;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Iterator;

import static java.util.Arrays.asList;

class BuildTypes extends TeamCityObject implements Iterable<BuildType> {

    @SerializedName("buildType")
    private final Collection<BuildType> buildTypes;

    public BuildTypes(BuildType... buildTypes) {
        this.buildTypes = new ArrayList<BuildType>(asList(buildTypes));
    }

    @Override
    public Iterator<BuildType> iterator() {
        return buildTypes.iterator();
    }
}

File: src/main/scala/bad/robot/radiate/teamcity/BuildTypes.scala
package bad.robot.radiate.teamcity

import argonaut.DecodeJson

object BuildTypes {
  implicit def buildTypesDecoder: DecodeJson[BuildTypes] = {
    DecodeJson(cursor => {
      (cursor --\ "buildType").as[List[BuildType]] map { BuildTypes(_) }
    })
  }
}

case class BuildTypes(list: List[BuildType]) extends TeamCityObject with Iterable[BuildType] {
  def iterator: Iterator[BuildType] = list.iterator
}